File: app/src/main/java/es/osw/cafelatte/domain/interactor/RxLoginUserInteractor.java
package es.osw.cafelatte.domain.interactor;

import es.osw.cafelatte.domain.model.LoginUser;
import es.osw.cafelatte.domain.model.User;
import es.osw.cafelatte.presentation.di.base.PerActivity;
import io.reactivex.Observable;
import javax.inject.Inject;
import timber.log.Timber;

@PerActivity public class RxLoginUserInteractor {
  @Inject public RxLoginUserInteractor() {
  }

  public Observable<User> createObservable(LoginUser param) {
    return Observable.just(createFakeUser())
        .doOnSubscribe(disposable -> Timber.d("Singing in %1$s in our servers", param.getEmail()));
  }

  private User createFakeUser() {
    return User.newBuilder().setEmail("email").setImage("image").setName("name").build();
  }
}

File: app/src/main/java/es/osw/cafelatte/domain/interactor/RxLoginUserInteractor.kt
package es.osw.cafelatte.domain.interactor

import com.google.firebase.auth.FirebaseUser
import es.osw.cafelatte.domain.model.Profile
import es.osw.cafelatte.domain.repository.UserRepository
import es.osw.cafelatte.presentation.di.base.PerActivity
import es.osw.cafelatte.presentation.model.mapper.ProfileMapper
import io.reactivex.Completable
import javax.inject.Inject

@PerActivity class RxLoginUserInteractor @Inject constructor(val userRepository: UserRepository) {

    fun createObservable(param: FirebaseUser?): Completable {
        val profile = ProfileMapper.map(param)
        return userRepository.insert(profile)
    }

    private fun createFakeUser(): Profile {
        return Profile("name", "email", "image")
    }
}
